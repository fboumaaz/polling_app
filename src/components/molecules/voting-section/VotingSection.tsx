import React, { FC, useState } from "react";
import {
  RadioGroup, FormControlLabel, Radio,
  Button, Typography, Grid, Box, Paper
} from "@mui/material";
import optionsProps from "../../shared/interfaces/OptionsProps";

interface Props {
  question: string;
  options: Array<optionsProps>;
  setOptions: (optionsProps: Array<optionsProps>) => void;
}

const VotingSection: FC<Props> = ({
                                    question,
                                    options,
                                    setOptions
                                  }: Props): JSX.Element => {
  const [selectedOption, setSelectedOption] = useState('');
  const OptionsList = (
    options.map((option) => (
      <FormControlLabel
        value={option.id}
        control={<Radio />}
        onChange={() => setSelectedOption(option.id)}
        label={option.answer}
        key={option.id}
      />
    ))
  );

  const vote = (() => {
    options.map((option) => {
      if (option.id === selectedOption) option.vote += 1;
    });
    setOptions([...options]);
  });

  return (
    <Paper variant="outlined" sx={{ flexGrow: 1, my: 2, p: 1.5 }}>
      <Grid rowSpacing={8} container>
        <Grid item sm={12}>
          <Typography
            variant="h5"
            color="secondary"
            align="center"
            sx={{ mb: 4 }}
            gutterBottom
          >
            Voting section
          </Typography>
          <Typography gutterBottom variant="h6" color="info">
            {question}
          </Typography>
          <RadioGroup>
            {OptionsList}
          </RadioGroup>
        </Grid>
        <Grid item sm={12}>
          <Box display="flex" justifyContent="flex-end">
            <Button
              color="secondary"
              variant="contained"
              sx={{ mb: 1, ml: 1, p: 0.3, mr: 1 }}
              disabled={options.length < 2}
              onClick={() => vote()}
              data-testid="voting-button"
            >
              Vote
            </Button>
          </Box>
        </Grid>
      </Grid>
    </Paper>
  );
};

export default VotingSection;
