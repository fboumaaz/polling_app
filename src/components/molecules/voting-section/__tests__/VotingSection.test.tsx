import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import VotingSection from "../VotingSection";
import { Button, RadioGroup, Typography } from "@mui/material";
import optionsData from "../../../shared/tests/options.data";

configure({ adapter: new Adapter() });

describe("VotingSection component", () => {
  const setOptions = jest.fn();

  describe("content", () => {
    const wrapper = shallow(<VotingSection options={[]} question={""} setOptions={setOptions} />);

    it("has a title", () => {
      expect(wrapper.find(Typography).at(0).text()).toBe("Voting section");
    });

    it("renders snapshots", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("has a voting button", () => {
      expect(wrapper.find(Button)).toHaveLength(1);
    });
  });

  describe("when options and question are empty", () => {
    const wrapper = shallow(<VotingSection options={[]} question={""} setOptions={setOptions} />);

    it("has a voting button disabled", () => {
      expect(wrapper.find(Button).prop("disabled")).toBe(true);
    });

    it("has empty list", () => {
      expect(wrapper.find(RadioGroup)).toHaveLength(1);
      expect(wrapper.find(RadioGroup).children()).toHaveLength(0);
    });
  });

  describe("when options and question are filled", () => {
    const wrapper = shallow(<VotingSection options={optionsData} question={"what is the capital of amsterdam?"}
                                           setOptions={setOptions} />);

    it("has 3 list items", () => {
      expect(wrapper.find(RadioGroup)).toHaveLength(1);
      expect(wrapper.find(RadioGroup).children()).toHaveLength(3);
    });

    it("has a voting button enabled", () => {
      expect(wrapper.find(Button).prop("disabled")).toBe(false);
    });

    it("calls setOptions when the button is clicked", () => {
      expect(setOptions).not.toHaveBeenCalled();
      wrapper.find(Button).simulate("click");
      expect(setOptions).toHaveBeenCalled();
    });
  });
});
