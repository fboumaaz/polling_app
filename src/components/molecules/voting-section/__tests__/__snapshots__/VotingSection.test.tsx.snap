// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VotingSection component content renders snapshots 1`] = `ShallowWrapper {}`;

exports[`VotingSection component renders snapshots 1`] = `ShallowWrapper {}`;

exports[`VotingSection component when options and question are filled renders snapshots 1`] = `ShallowWrapper {}`;
